File: 00xx/000x/0001-TwoSum/Solution.java
/*  Solution: Hash Map
    Time complexity:    O(n)
    Space complexity:   O(n)
 */

class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<>();
        int complement;
        for (int i = 0; i < nums.length; i++) {
            complement = target - nums[i];
            if (map.containsKey(complement)) {
                return new int[] {map.get(complement), i};
            }
            else {
                map.put(nums[i], i);
            }
        }
        return new int[] {};
    }
}

File: 00xx/000x/0001-TwoSum/Solution.cc
/*  Solution: Brute Force
    Time Complexity:    O(n^2)
    Space Complexity:   O(1)

    Note: This is a pretty obvious but mediocre solution.
 */

class Solution {
public:
  vector<int> twoSum(vector<int>& nums, int target) {
    for (int i = 0; i < nums.size(); i++) {
      for (int j = i + 1; j < nums.size(); j++) {
        if (nums[i] + nums[j] == target) {
          return vector<int>{i, j};
        }
      }
    }
    return vector<int>{};
  }
};
